### ISSUE TITLE
Remove candidates from the completion system after action

### ISSUE THREAD

**Issue by minad** [oantolin/embark#117]
Is it possible to remove candidates from the completion system after executing a delete action in particular in combination with embark-act-noexit, which I favor greatly over embark-act. This is a relatively minor thing and only applies to delete operations. In the case of static completion tables it does not suffice to refresh the completion system. Probably it is not worth to solve this inside Embark, but maybe you have given it already some thought. One would probably need another completion system `embark-delete-candidate-hook` and then register the commands somehow as deletion commands.

**Comment by oantolin** [comment]
If refreshing the completion system is not enough it sounds more like the completion table is broken? Why is it offering candidates that have been deleted? As you say one way this can happen is that the completion table is mistakenly static and does not check if it's candidates still make sense. I would argue that in that case being static is a bug in the completion table being used.

**Comment by minad** [comment]
The point is, before adding Embark to the mix, the completion table is really static like for example the list of bookmarks. Only with embark-act-noexit, this does not hold anymore since now you have actions manipulating the table while the completion is still ongoing. Therefore I argue that there is really nothing wrong at all with the static completion tables in such cases. This obviously only holds if the list is not manipulated in place, then the update may even work with static completion tables. But this is fragile when the head is mutated itself or a new element is prepended.

**Comment by oantolin** [comment]
I take it back partially. I think completion tables are allowed to generate nonsense candidates that are later filtered or t by the completion predicate. So in the case you mention maybe we should consider the problem not as a bug in the static completion table, but a bug in the completion predicate.

**Comment by minad** [comment]
No, I think there is really no bug - in neither the predicate nor the table. The completion table can not foresee that it will be magically manipulated afterwards by Embark. I think you should take back your entire statement ;)

**Comment by oantolin** [comment]
> Only with embark-act-noexit, this does not hold anymore since now you have actions manipulating the table while the completion is still ongoing

There are lots of ways a candidate can become invalid during a completion session without using embark. You can have recursive minibuffers and just run a command that deletes something, or you could switch away from the minibuffer, kill some buffer and switch back to the minibuffer.

**Comment by oantolin** [comment]
Embark just runs commands, so any problem you think Embark is causing here can also be triggered without embark and not in some crazy convoluted way, but just a little less conveniently than with embark.

**Comment by oantolin** [comment]
And its not just deleting candidates, new candidates should maybe appear if you rename a buffer, or create a new file, etc. Again all these operations can be done with or without embark. I think many completion sessions are just not made for being kept open while you do a bunch of other stuff. Even so many builtin completion tables do the right thing, and simply refreshing the UI is enough. I think the completion tables of find-file and switch-to-buffer are good examples, they seem to always know the correct candidate list.

**Comment by oantolin** [comment]
So I agree that stake candidate lists are an annoying problem but I disagree that embark causes them. Maybe you could say embark makes it more convenient for the user to trigger this problem. I'm OK with that, Embark is meant to let you do stuff you could do anyway in a more convenient fashion. 😛

**Comment by oantolin** [comment]
That was meant to say "stale candidate lists", GitHub is not letting me edit.

**Comment by minad** [comment]
> There are lots of ways a candidate can become invalid during a completion session without using embark. You can have recursive minibuffers and just run a command that deletes something, or you could switch away from the minibuffer, kill some buffer and switch back to the minibuffer.

Sure, but this is not the fault of the static completion table. It is okay if a completion table is static in such cases. It must not prepare for being manipulated by external side effects. While I usually try to make things robust with recursive minibuffers, I really do think that it is okay like this. Otherwise almost every completion table would have to be dynamic, except the ones which are lexically bound constant variables, e.g., `(completing-read "Prompt: " '(my really constant candidates)`. This is just not a practical way of looking at it. I think it is more practical to take the point of view that candidate tables can be considered static if they can only be modified by external recursively triggered side effects. Embark belongs to that class of side effects. Note that this seems also to be the Emacs interpretation of things. For example bookmark-jump does not work if I call embark-act-noexit and bookmark-delete.

But if you say that this is out of scope of Embark that is also fine - I just thought about pointing this out. Another problematic operation is renaming.

**Comment by minad** [comment]
> Embark just runs commands and inputs the target for you, so any problem you think Embark is causing here can also be triggered without embark and not in some crazy convoluted way, but just a little less conveniently than with embark.

Agree. But it is a common thing to have an Embark delete action.

> So I agree that stake candidate lists are an annoying problem but I disagree that embark causes them. Maybe you could say embark makes it more convenient for the user to trigger this problem. I'm OK with that, Embark is meant to let you do stuff you could do anyway in a more convenient fashion.

Agree. It is not really Embark causing this. This is not a bug. But Embark could do something against this problem - such that Embark acquires a new feature if you deem this to be worth it ;)

**Comment by oantolin** [comment]
> almost every completion table would have to be dynamic

That's the right answer, of course, but let's see if we can't fake it somehow. I wouldn't want to special-case deletions, but would prefer to implement a solution that also handles renaming, etc.

How about just restarting the command you are running and injecting the same input you have?

```elisp
(defun embark-refresh ()
  "Restart current command with current input.
Use this to refresh the list of candidates for commands that do
not handle that themselves."
  (interactive)
  (when (minibufferp)
    (let ((target (minibuffer-contents))
          (command embark--command))
      (run-at-time 0 nil (lambda ()
                           (minibuffer-with-setup-hook
                               (lambda ()
                                 (delete-minibuffer-contents)
                                 (insert target))
                             (let ((use-dialog-box nil)
                                   (this-command command))
                               (command-execute command)))))
      (abort-recursive-edit))))
```

That can be called interactively or put in `embark-post-action-hook`.

**Comment by minad** [comment]
> That's the right answer, of course, but let's see if we can't fake it somehow. I wouldn't want to special-case deletions, but would prefer to implement a solution that also handles renaming, etc.

I fully agree. This would be much better. I will try the refresh/restart idea to check how well it works! Maybe call it embark-restart instead?

**Comment by oantolin** [comment]
To be honest, I only tried running it directly, not from the post action hook. I'll investigate.

**Comment by minad** [comment]
I think one has to check for active-minibuffer-window and switch there.

**Comment by oantolin** [comment]
Ah, the action runs in the target window, and thus the post action hook does too.

**Comment by oantolin** [comment]
Yeah, that seems to do it:

```elisp
(defun embark-refresh ()
  "Restart current command with current input.
Use this to refresh the list of candidates for commands that do
not handle that themselves."
  (interactive)
  (when-let ((miniwin (active-minibuffer-window)))
    (with-selected-window miniwin
      (let ((target (minibuffer-contents))
            (command embark--command))
        (run-at-time 0 nil (lambda ()
                             (minibuffer-with-setup-hook
                                 (lambda ()
                                   (delete-minibuffer-contents)
                                   (insert target))
                               (let ((use-dialog-box nil)
                                     (this-command command))
                                 (command-execute command)))))
        (abort-recursive-edit)))))
```

**Comment by oantolin** [comment]
> Maybe call it embark-restart instead?

I was already considering `embark-restart` and `embark-revert`. Maybe revert is my favorite.

Some notes:

This function is very similar to `embark-become`. The differences are (1) the obvious one: it uses `embark--command` instead of prompting you, (2) it uses the whole minibuffer contents, while `embark-become` uses only the portion between the current completion boundaries, (3) it uses `abort-recursive-edit` instead of `top-level`.

For (2), embark-become uses just the part between the boundaries mainly so that if you switch from, say `find-file` to `switch-to-buffer`, you only get the last part of the path in the `switch-to-buffer` prompt (the full path would basically never match a buffer name).

About (3): it was decided long ago that "exiting" in embark meant `(top-level)` and I've often wondered if it should exit just one level of minibuffer...

**Comment by minad** [comment]
embark-restart and embark-revert are both good names. I like embark-restart a bit more, but don't feel strongly about it.

Regarding (3), I would like it more if Embark exits only one level. I didn't really know this but I somehow assumed that Embark exits one level. But as you know, I am usually calling embark-act-noexit.

There is one more issue with embark-restart - if you are using the arrow keys to select a candidate in Selectrum, the current selection will be lost during the restart. In order to fix this one either needs a small advice for embark-restart or maybe an additional hook.

**Comment by oantolin** [comment]
`embark-restart` it is, then.

For exiting, I'm tempted to just change the exit to be one level only without mentioning it. People who don't use recursive minibuffers will not even notice, and people who do use recursive minibuffers will probably all prefer exiting just one level.

I think I'll add two hooks to embark restart, one to save any state you need to save, and one where you restore it.

**Comment by minad** [comment]
> I think I'll add two hooks to embark restart, one to save any state you need to save, and one where you restore it.

I would add a single save hook which returns a lambda to restore.

**Comment by oantolin** [comment]
> I would add a single save hook which returns a lambda to restore.

Good idea, I'll do that.

**Comment by minad** [comment]
There is one minor related issue regarding the refreshing behavior. When executing actions in Selectrum the currently selected candidate is lost and the first candidate is selected again. This happens in consult-buffer and also in other commands. I would like to walk over the list, execute actions on the go. For my uses cases the only "serious issues" I have with Embark is this issue #117 and the which-key submenu issue #139.

**Comment by oantolin** [comment]
For me, this issue is mostly caused by Selectrum and Consult's static nature. I mean, it is an issue that does affect default completion and Icomplete and some built-in Emacs commands, *but* the most common cases there are fully dynamic: if you use `find-file` or `switch-to-buffer` and while the minibuffer is open you add or delete files or buffers, as soon as you refresh Icomplete you get the new list!

So for the most part I see this as a cache invalidation bug in Selectrum and in Consult's candidate lists. But I think that because of other benefits, it's a bug worth keeping in Selectrum and Consult, and just trying to add a work around here. 😉 Note that Embark does not contribute at all to this problem: if you enable recursive minibuffers and add or delete files or buffers without using Embark at all, Selectrum still exhibits this cache invalidation bug.

So, in summary: Embark has nothing to with this, it's a bug in Selectrum and `consult-buffer`, but Embark is happy to offer a workaround. 😛

By the way, the reason I didn't push the `embark-refresh` command yet is that I wanted to make sure I could get Selectrum to keep the candidate index first, but I couldn't quite figure it out. But this should be easy. I'll dust off the version of the function that did have the hook (I see the one above doesn't) and try again or ask for help.

**Comment by minad** [comment]
Haha, I still think it is a bug in Embark. There are probably many cases floating around where you have `(completing-read "prompt: " some-list)` - as I argued all of these cases will not work when the head of the list is removed/replaced. It will work if you mutate the tail of the list somewhere. This is a weird distinction, we should not mutate lists :-P

> By the way, the reason I didn't push the embark-refresh command yet is that I wanted to make sure I could get Selectrum to keep the candidate index first, but I couldn't quite figure it out. But this should be easy. I'll dust off the version of the function that did have the hook (I see the one above doesn't) and try again or ask for help.

Sure, no hurry. It shouldn't be a big problem and I hope @clemera can help with that if some things don't work as they should.

**Comment by oantolin** [comment]
> Haha, I still think it is a bug in Embark.

How can it be Embark's fault when I can reproduce it with Selectrum and `enable-recursive-minibuffer` without even loading Embark? 😛

**Comment by oantolin** [comment]
> There are probably many cases floating around where you have `(completing-read "prompt: " some-list)` - as I argued all of these cases will not work when the head of the list is removed/replaced. It will work if you mutate the tail of the list somewhere. This is a weird distinction, we should not mutate lists :-P

But this is true. (And again, unrelated to Embark. :D) I think Emacs has the right idea: make the important and common cases, namely files and buffer, use fully dynamic completion tables. In other words treat passing a list to `completing-read` as a declaration that you agree with static behavior.

**Comment by oantolin** [comment]
> How can it be Embark's fault when I can reproduce it with Selectrum and `enable-recursive-minibuffers` without even loading Embark? 😛

For files I don't even need recursive minibuffers: just a separate terminal. 😉

**Comment by minad** [comment]
> How can it be Embark's fault when I can reproduce it with Selectrum and enable-recursive-minibuffers without even loading Embark?

Enable-recursive-minibuffers is disabled by default in order to prevent you from breaking out :)

> In other words treat passing a list to completing-read as a declaration that you agree with static behavior.

Yes, but this is the point where Embark comes into play and makes static things dynamic. This is something the original author didn't even anticipate (modulo the recursive minibuffers, but these are disabled).

> For files I don't even need recursive minibuffers: just a separate terminal.

Well, this does not exist really. There is only the Emacs. You should install EXWM and keep recursive minibuffers disabled in order to prevent you from doing weird things ;)

**Comment by minad** [comment]
Current status - with Vertico the following two approaches seem to work.

1. `embark-refresh` by @oantolin, see https://github.com/oantolin/embark/issues/117#issuecomment-762901648

~~~ elisp
(defun embark-refresh ()
  "Restart current command with current input.
Use this to refresh the list of candidates for commands that do
not handle that themselves."
  (interactive)
  (when-let ((miniwin (active-minibuffer-window)))
    (with-selected-window miniwin
      (let ((target (minibuffer-contents))
            (command embark--command))
        (run-at-time 0 nil (lambda ()
                             (minibuffer-with-setup-hook
                                 (lambda ()
                                   (delete-minibuffer-contents)
                                   (insert target))
                               (let ((use-dialog-box nil)
                                     (this-command command))
                                 (command-execute command)))))
        (abort-recursive-edit)))))

(setq embark-quit-after-action nil)
(add-hook 'embark-post-action-hook #'embark-refresh)
~~~

2. Using `vertico-repeat`

~~~ elisp
(setq embark-quit-after-action t)
(add-hook 'embark-pre-action-hook
          (lambda ()
            (setq-default vertico-repeat--restore t)))
(add-hook 'embark-post-action-hook
          (lambda ()
            (setq-default vertico-repeat--restore nil
                          vertico-repeat--input "")
            (vertico-repeat)))
~~~

**Comment by jdtsmith** [comment]
Great discussion/potential solution, especially the analogy with embark-become.  I agree with "it's not Embark's fault, but it exposes the problem to many more users, and has an opportunity to help."  

It's too bad there is not some simple event system in place for mini-buffer displayers, whomever they are, to receive the message "I have potentially invalidated your list of results; refresh yourselves accordingly."

**Comment by minad** [comment]
Yes, I thought about proposing the addition of such completion UI hooks for refreshing. See also the consult ui integration code, e.g., consult-vertico.

**Comment by oantolin** [comment]
I don't remember very well now, but on the Wiki there was some code to refresh icomplete or embark-collect after an action. That approach should probably work with Vertico too. I'll take a look when I get the chance (which won't be for a couple of hours at least).

**Comment by jdtsmith** [comment]
@minad: for the `vertico-repeat` approach, shouldn't that run only if embark is _not_ exiting?

**Comment by oantolin** [comment]
For Vertico, the following seems to work: `(add-hook 'embark-post-action-hook #'consult-vertico--refresh)`. (Of course, you need to install Consult for that function.)

**Comment by minad** [comment]
@oantolin No, this is not a good recommendation as per the discussion above. Static completion lists will not work with consult-vertico--refresh.

**Comment by oantolin** [comment]
Static completions lists can't change, unless I'm confused about what static means. This question is about dynamic completion tables, isn't it?

**Comment by oantolin** [comment]
`consult-vertico--refresh` seems to work fine with the file and buffer completion tables built-in to Emacs... Oh, are you saying it won't work with, say, `consult-buffer`?

**Comment by minad** [comment]
Fixed in #326.

**Comment by jdtsmith** [comment]
Works great, thanks.
